spatialos-sdk/src/worker/schema.rs

```diff
@@ -41,3 +109,151 @@ pub trait SchemaPrimitiveField {
+    }
+}
+
+/// The error type returned when a conversion to a schema enum type fails.
+#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct UnknownDiscriminantError;
+
+/// A type that corresponds to a [schemalang enumeration][schema-enum].
+///
+/// Schemalang enums are C-like enums (i.e. they only have a discriminant and do not
+/// contain any additional data) represented as `u32` values when serialized. They
+/// default to the first variant defined in their schema declaration.
+///
+/// Converting the serialized `u32` representation back to the Rust representation
+/// can fail if the serialized discriminant doesn't match any of the recognized
+/// variants (e.g. if the serialized data was generated from a different schema
+/// version than what the current program is using). As such, conversion is handled
+/// using the [`TryFrom`] trait, with [`UnknownDiscriminantError`] indicating that
+/// value couldn't be converted.
+///
+/// You should generally not have to implement this trait manually for any types.
+/// The implementation for any schema-defined types will be generated for you by the
+/// code generator provided with the SDK.
+///
+/// [schema-enum]: https://docs.improbable.io/reference/14.2/shared/schema/reference#enumerations
+/// [`TryFrom`]: https://doc.rust-lang.org/std/convert/trait.TryFrom.html
+/// [`UnknownDiscriminantError`]: struct.UnknownDiscriminantError.html
+pub trait EnumField:
+    Sized + Clone + Copy + Default + TryFrom<u32, Error = UnknownDiscriminantError> + Into<u32>
+{
+}
+
+/// Helper macro for generating the [`Field`] implementation for a schema enum.
+///
+/// Ideally we would like to provide a blanket impl of [`Field`] for all types
+/// implementing [`EnumField`] the way we do with [`ObjectField`], but Rust's
+/// current coherence rules mean that there can only be one blanket impl for a given
+/// trait. For now, this macro at least provides a lightweight way to implement the
+/// trait for such types. Once the coherence rules [have been made more flexible][rfc-1053],
+/// we can remove this macro and use a regular blanket impl.
+///
+/// You should generally not have to use this macro manually. The code generated for
+/// schema-defined enums will handle this for you.
+///
+/// [`Field`]: trait.Field.html
+/// [`EnumField`]: trait.EnumField.html
+/// [`ObjectField`]: trait.ObjectField.html
+/// [rfc-1053]: https://github.com/rust-lang/rfcs/issues/1053
+#[macro_export]
+macro_rules! impl_field_for_enum_field {
+    ($type:ty) => {
+        impl $crate::worker::schema::Field for $type {
+            type RustType = Self;
+
+            fn get_or_default(
+                object: &$crate::worker::schema::SchemaObject,
+                field: $crate::worker::schema::FieldId,
+            ) -> Self::RustType {
+                Self::try_from(object.get::<$crate::worker::schema::SchemaEnum>(field))
+                    .unwrap_or_default()
```

Comment:
It would be nice to be able to not lose the fact that there was an error here.

Reply:
So, per my previous comment, I had assumed that schema enums can reasonably default to their first variant, which many not be a valid assumption. Is there any concept of defaults for enums in schemalang? If not, handling this case may actually necessitate that deserialization return a `Result` 😲

Reply:
We finally found the use case where deserialization can fail 🙃 

Its always enums that break things :(

Reply:
I'd like to do that in a separate PR if you're cool with that. It's going to be a fairly noisy change, and there are some other error cases that I was previously ignoring that I'd like to add handling for at the same time. For now, I can either leave the logic for providing a default value (which may still be reasonable to do, per @dgavedissian's suggestion), or I can have it panic if value is invalid for the enum.

Reply:
Lets leave the current behaviour and revisit in a future PR 😄